sasl/digest/src/main/java/org/wildfly/security/sasl/digest/AbstractDigestMechanism.java

```diff
@@ -214,169 +214,168 @@ protected DigestWrapper(boolean confidential) {
+                System.arraycopy(clearText, 0, decryptedMessage, 0, clearText.length - 10);
+            }
 
-        if (messageType != 1) {
-            throw saslDigest.mechMessageTypeMustEqual(1, messageType).toSaslException();
-        }
+            byte[] expectedHmac = computeHMAC(unwrapHmacKeyIntegrity, extractedSeqNum, hmacMD5, decryptedMessage, 0, decryptedMessage.length);
 
-        if (extractedSeqNum != unwrapSeqNum) {
-            throw saslDigest.mechBadSequenceNumberWhileUnwrapping(unwrapSeqNum, extractedSeqNum).toSaslException();
-        }
+            // check hmac-s
+            if (Arrays2.equals(expectedHmac, 0, hmac, 0, 10) == false) {
+                return NO_BYTES;
+            }
 
-        byte[] clearText = null;
-        try {
-            clearText = unwrapCipher.update(message, offset, len - 6);
-        } catch (Exception e) {
-            throw saslDigest.mechProblemDuringDecrypt(e).toSaslException();
-        }
-        if (clearText == null){
-            throw saslDigest.mechProblemDuringDecryptResultIsNull().toSaslException();
+            unwrapSeqNum++; // increment only if MAC is valid
+            return decryptedMessage;
         }
-
-        byte[] hmac = new byte[10];
-        System.arraycopy(clearText, clearText.length - 10, hmac, 0, 10);
-
-        byte[] decryptedMessage = null;
-        // strip potential padding
-        if (unwrapCipher.getBlockSize() > 0) {
-            int padSize = clearText[clearText.length - 10 - 1];
-            int decryptedMessageSize = clearText.length - 10;
-            if (padSize < 8) {
-                int i = clearText.length - 10 - 1;
-                while (clearText[i] == padSize) {
-                    i--;
-                }
-                decryptedMessageSize = i + 1;
+        @Override
+        public byte[] wrap(byte[] outgoing, int offset, int len) throws SaslException {
+            if (confidential) {
+                return wrapConfidentialityProtectedMessage(outgoing, offset, len);
+            } else {
+                return wrapIntegrityProtectedMessage(outgoing, offset, len);
             }
-            decryptedMessage = new byte[decryptedMessageSize];
-            System.arraycopy(clearText, 0, decryptedMessage, 0, decryptedMessageSize);
-        } else {
-            decryptedMessage = new byte[clearText.length - 10];
-            System.arraycopy(clearText, 0, decryptedMessage, 0, clearText.length - 10);
         }
 
-        byte[] expectedHmac = computeHMAC(unwrapHmacKeyIntegrity, extractedSeqNum, hmacMD5, decryptedMessage, 0, decryptedMessage.length);
-
-        // check hmac-s
-        if (Arrays2.equals(expectedHmac, 0, hmac, 0, 10) == false) {
-            return NO_BYTES;
+        /* (non-Javadoc)
+         * @see org.wildfly.security.sasl.util.SaslWrapper#unwrap(byte[], int, int)
+         */
+        @Override
+        public byte[] unwrap(byte[] incoming, int offset, int len) throws SaslException {
```

Comment:
@keshav-redhat Thank you for the PR! Can you please put the private methods you are moving under the public methods? Just a convention, it makes it easier to read and it separates API from implementation

Reply:
@keshav-redhat  also, try to not make unrelated changes in this file so that this PR is not so cluttered